include/mfmg/dealii_operator_device.templates.cuh

```diff
@@ -145,6 +147,117 @@ SparseMatrixDeviceOperator<VectorType>::build_range_vector() const
 
 //-------------------------------------------------------------------------//
 
+namespace internal
+{
+template <typename ScalarType>
+__global__ void extract_inv_diag(ScalarType const *const matrix_value,
+                                 int const *const matrix_column_index,
+                                 int const *const matrix_row_index,
+                                 int const size, ScalarType *value)
+{
+  int idx = threadIdx.x + blockIdx.x * blockDim.x;
+  if (idx < size)
+    if (matrix_column_index[idx] == matrix_row_index[idx])
+      value[matrix_column_index[idx]] = 1. / matrix_value[idx];
+}
+}
+
+template <typename VectorType>
+SmootherDeviceOperator<VectorType>::SmootherDeviceOperator(
+    matrix_type const &matrix,
+    std::shared_ptr<boost::property_tree::ptree> params)
+    : _matrix(matrix)
+{
+  std::string prec_type = params->get("smoother.type", "Jacobi");
+  initialize(prec_type);
+}
+
+template <typename VectorType>
+void SmootherDeviceOperator<VectorType>::apply(VectorType const &b,
+                                               VectorType &x) const
+{
+  // r = -(b - Ax)
+  vector_type r(b);
+  _matrix.vmult(r, x);
+  r.add(-1., b);
```

Comment:
Maybe implement an proper `alpha*A*x + beta*y` `apply` call for the device matrix? Then, this can be done more efficiently.

Reply:
Let me open an issue for this because I plan to switch later to the deal.II distributed vectors. So I don't want to implement something that we won't use for a long time.